From 038c0b625c504ade8ea0ecc669c8a4d8c99dd25c Mon Sep 17 00:00:00 2001
From: Michael Braun <michael-dev@fami-braun.de>
Date: Mon, 11 May 2015 11:00:57 +0200
Subject: [PATCH] refuse incoming packets with local source mac

sometimes bridge code gets confused by local wifi packets getting back at us, which will then fail packet delivery.
This tries to fix this by filtering this out in tinc.

Symptoms: bridge fdb dead, ipv6 warning
---
 src/route.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/route.c b/src/route.c
index 5782c56..9f1466a 100644
--- a/src/route.c
+++ b/src/route.c
@@ -834,6 +834,15 @@ static void route_mac(node_t *source, vpn_packet_t *packet) {
 		mac_t src;
 		memcpy(&src, &packet->data[6], sizeof src);
 		learn_mac(&src, vlan_id);
+	} else {
+		/* refuse packets from others with source mac owned by us */
+		mac_t src;
+		memcpy(&src, &packet->data[6], sizeof src);
+		subnet = lookup_subnet_mac(myself, &src, vlan_id);
+		if(subnet && subnet->owner == myself) {
+			ifdebug(TRAFFIC) logger(LOG_WARNING, "Packet looping back to us from %s (%s)!", source->name, source->hostname);
+			return;
+		}
 	}
 
 	/* Lookup destination address */
-- 
1.9.1

